node/src/components/contract_runtime/operations.rs

```diff
@@ -23,13 +25,52 @@ use crate::{
             BlockAndExecutionEffects, ExecutionPreState, Metrics,
         },
     },
-    types::{Block, Deploy, DeployHeader, FinalizedBlock},
+    types::{ActivationPoint, Block, Deploy, DeployHeader, FinalizedBlock},
 };
 use casper_execution_engine::{
     core::{engine_state::execution_result::ExecutionResults, execution},
     storage::global_state::{CommitProvider, StateProvider},
 };
 
+fn generate_range_by_index(n: u64, m: u64, index: u64) -> Option<Range<u64>> {
+    let start = index.checked_mul(m)?;
+    let end = cmp::min(start.checked_add(m)?, n);
+    Some(start..end)
+}
+
+/// Calculates era keys to be pruned.
+///
+/// Outcomes:
+/// * Ok(Some(range)) -- these keys should be purged
+/// * Ok(None) -- nothing to do, either done, or there is not enough eras to purge
+fn calculate_purge_eras(
+    activation_era_id: EraId,
+    activation_height: u64,
+    current_height: u64,
+    batch_size: u64,
+) -> Option<Vec<Key>> {
+    if batch_size == 0 {
+        // Nothing to do, the batch size is 0.
+        return None;
+    }
+
+    let nth_chunk: u64 = match current_height.checked_sub(activation_height) {
+        Some(nth_chunk) => nth_chunk,
+        None => {
+            // Time went backwards, programmer error, etc
```

Comment:
Seems like we should at least log an error here.

Reply:
As discussed offline (and noted in your own comment below) we agreed to change this to panicking code, e.g. `current_height.checked_sub(activation_height).expect("...")`

Reply:
My bad, I didn't check in this change. Added a `panic!` and changed `warn!` to `error!`,